### PR TITLE
ruby: vendor gem like we do pip for Python

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

CC @ilovezfs. PoC for the discussion at the end of https://github.com/Homebrew/homebrew-core/pull/19475.

### CODE REVIEW

**Comment by DomT4** [comment]
Failure seems unrelated. Reproduces with & without this PR.

File: Formula/ruby.rb [hunk 1]
```diff
@@ -32,6 +33,23 @@ class Ruby < Formula
...
+
+  def program_suffix
+    build.with?("suffix") ? "24" : ""
+  end
+
+  def ruby
+    "#{bin}/ruby#{program_suffix}"
+  end
+
+  def abi_version
```
Comment (JCount) [inline Formula/ruby.rb]: Shouldn't this be `api_version`?

Comment (DomT4) [reply Formula/ruby.rb]: Umm. I _think_ so, but considerably more intelligent & experienced devs than me have touched this code quite a bit over the years & left it as-is so I'm reluctant to disagree with them to an extent 🙈.

Comment (JCount) [reply Formula/ruby.rb]: `#{ruby} -e 'print Gem.ruby_api_version'` 🤔

Comment (DomT4) [reply Formula/ruby.rb]: Yeah, I got the logic. Just... find it amusingly difficult to believe we've been writing it this way for years and... nobody has noticed? 🙈. I mean it's possible, but 😅.

**Comment by ilovezfs** [comment]
@xu-cheng any thoughts here?

**Comment by DomT4** [comment]
If people are a 👍 on the idea I'll roll it out across the whole set of Ruby formulae, but keeping it at just the main one for now to avoid repeated large CI jobs.

**Review by MikeMcQuaid — APPROVED** [review]
This seems like a good idea, nice work @DomT4 and 👍 to doing it on more formulae.

**Comment by DomT4** [comment]
TIL that we were already vendoring an older Rubygems for `ruby@1.8`. Wish I'd checked that before investing an hour+ into testing my own method the other night 🤣. Anyway, expanded my method across all the Rubies so we'll see if CI explodes at all.

**Comment by DomT4** [comment]
Rebased against master to resolve the merge conflict.

**Comment by DomT4** [comment]
Rebased again because CI fell apart last night on a whole variety of builds, including this one.

**Comment by ilovezfs** [comment]
:gem:

**Comment by DomT4** [comment]
Thanks for the discussion & review all :heart:.

**Pull request closed without merging**